packages/flutter_tools/test/integration.shard/test_flutter_tree.dart

```diff
@@ -0,0 +1,391 @@
+    ensureToolWithFakeDart();
+    assert(flutterToolsStampFile.readStringLikeShell() == stampValue);
+
+    _warmTree = fileSystem
+      .systemTempDirectory.createTempSync('flutter_test_tree_warm.').absolute;
+    _rsyncTreesSync(root, _warmTree!);
+  }
+
+  void _dispose() {
+    try {
+      root.deleteSync(recursive: true);
+    } on FileSystemException {
+      // ignore
+    }
+    try {
+      _warmTree?.deleteSync(recursive: true);
+    } on FileSystemException {
+      // ignore
+    }
+  }
+
+  File get fakeDartLog => binCacheDir.childFile('fake-dart.log'); // bin/cache/fake-dart.log
+  File get dartBinary => dartSdkDir.childDirectory('bin').childFile('dart'); // bin/cache/dart-sdk/bin/dart
+  File get dartBinaryOrig => dartSdkDir.childDirectory('bin').childFile('dart.orig'); // bin/cache/dart-sdk/bin/dart.orig
+
+  /// Run the tool entrypoint to update the cache, with the Dart binary faked out.
+  ///
+  /// The fake `dart` logs the commands it receives, and this method returns
+  /// the list of log entries, for testing the entrypoint's behavior.
+  ///
+  /// The fake `dart` ignores most possible commands, except for logging them.
+  /// For some commands used by the entrypoint when it tries to update the tool,
+  /// it provides the needed behavior using shortcuts for efficiency:
+  ///
+  ///  * A command that looks like the entrypoint's `dart pub upgrade`
+  ///    will fake it by updating the last-modified time on `pubspec.lock`.
+  ///    (Each `pubspec.lock` in the baseline tree comes from [hostFlutterTree],
+  ///    so effectively we rely on those being up to date.)
+  ///
+  ///  * A command that looks like the entrypoint's command to generate the
+  ///    tool snapshot will fake it by copying from [hostFlutterTree].
+  ///
+  /// In particular, if this tree contains changes relative to [hostFlutterTree]
+  /// that would affect the behavior of the tool, the resulting snapshot will
+  /// not reflect those changes.
+  List<String> ensureToolWithFakeDart() {
+    fakeDartLog.writeAsStringSync('');
+    dartBinary.renameSync(dartBinaryOrig.path);
+    _writeFakeDart();
+    ensureToolSync();
+    dartBinaryOrig.renameSync(dartBinary.path);
+    return fakeDartLog.readAsLinesSync();
+  }
+
+  /// Write to [dartBinary] a script that fakes out the `dart` command.
+  ///
+  /// See [ensureToolWithFakeDart].
+  void _writeFakeDart() {
+    dartBinary.writeAsStringSync('''
+#!/usr/bin/env bash
+
+full_command="dart \$*"
+
+function log_command() {
+  local description="\$1"
+  echo "\$description: \$full_command" >>${shellEscapeString(fakeDartLog.path)}
+}
+
+case "\$*" in
+  "pub upgrade "*)
+    # This is a `dart pub upgrade` command, as in pub_upgrade_with_retry .
+    # Just update the last-modified time on the pubspec.lock .
+    touch pubspec.lock
+    log_command "pub upgrade"
+    ;;
+
+  *" --disable-dart-dev "*" --snapshot-kind=app-jit "*)
```

Comment:
I worry that this will break when we eventually change the CLI args. Plus there's a lot of complexity here with `shellEscapeString` that could be potentially difficult to debug. Can we just invoke the real dart bash script and then check if the snapshot was rebuilt?

Reply:
(I added a comment in the `shellEscapeString` implementation, adapted from my explanation above: https://github.com/flutter/flutter/commit/d3a37c776662a8eb4769511b25e139ce11cf2b93)

Reply:
Sorry for the delayed response on this one. I appreciate all the careful work that went into writing up this test harness. However, I feel its complexity does not match value added by this feature. In particular, I'm uncomfortable with:

1. relying on rsync
2. manipulating git
3. encoding shell escape rules

What about a more minimal test, that validates munging `flutter_tools/pubspec.yaml` causes a rebuild?